### PR TITLE
SonicWall Firewall CEF Connector

### PR DESCRIPTION
**Pull request by tijuc**
## Proposed Changes

  - Added SonicWall to Azure Sentinel for CEF data connectors

### CODE REVIEW

File: DataConnectors/SonicWall/SonicwallFirewall.JSON [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    "connectivityCriterias": [
+        {
+            "type": "IsConnectedQuery",
+            "value": [
+                "CommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n| summarize LastLogReceived = max(TimeGenerated)\n| project IsConnected = LastLogReceived > ago(30d)"
+            ]
+        }
+    ],
+    "availability": {
+        "status": 1
```
Comment (v-jayakal) [inline DataConnectors/SonicWall/SonicwallFirewall.JSON:39]: Please add "isPreview": true, refer https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_CEF_template.json

File: DataConnectors/SonicWall/SonicwallFirewall.JSON [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                    "description": "Install the Microsoft Monitoring Agent on your Linux machine and configure the machine to listen on the necessary port and forward messages to your Azure Sentinel workspace. The CEF collector collects CEF messages on port 514 TCP.\n\n> 1. Make sure that you have Python on your machine using the following command: python -version.\n\n> 2. You must have elevated permissions (sudo) on your machine.",
+                    "instructions": [
+                        {
+                            "parameters": {
+                                "fillWith": [
+                                    "WorkspaceId",
+                                    "PrimaryKey"
+                                ],
+                                "label": "Run the following command to install and apply the CEF collector:",
+                                "value": "sudo wget https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/DataConnectors/CEF/cef_installer.py&&sudo python cef_installer.py {0} {1}"
```
Comment (v-jayakal) [inline DataConnectors/SonicWall/SonicwallFirewall.JSON:85]: Missing -O cef_installer.py from the command, refer the template https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_CEF_template.json

File: DataConnectors/SonicWall/SonicwallFirewall.JSON [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            "title": "3. Validate connection",
+            "description": "Follow the instructions to validate your connectivity:\n\nOpen Log Analytics to check if the logs are received using the CommonSecurityLog schema.\n\n>It may take about 20 minutes until the connection streams data to your workspace.\n\nIf the logs are not received, run the following connectivity validation script:\n\n> 1. Make sure that you have Python on your machine using the following command: python -version\n\n>2. You must have elevated permissions (sudo) on your machine",
+            "instructions": [
+                {
+                    "parameters": {
+                        "fillWith": [
+                            "WorkspaceId"
+                        ],
+                        "label": "Run the following command to validate your connectivity:",
+                        "value": "sudo wget https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/DataConnectors/CEF/cef_troubleshoot.py&&sudo python cef_troubleshoot.py  {0}"
```
Comment (v-jayakal) [inline DataConnectors/SonicWall/SonicwallFirewall.JSON:107]: Missing  -O cef_troubleshoot.py . Refer template.

File: DataConnectors/Templates/Doc_Template_CEF_Connector.md [hunk 4]
```diff
@@ -18,4 +18,4 @@ Configure <<Partner Appliance Name>> to forward Syslog messages in CEF format to
 In this document, you learned how to connect <<Partner Appliance Name>> to Azure Sentinel. To learn more about Azure Sentinel, see the following articles:
 - Learn how to [get visibility into your data, and potential threats](quickstart-get-visibility.md).
```
Comment (v-jayakal) [inline DataConnectors/Templates/Doc_Template_CEF_Connector.md:19]: Can you please revert the changes to this file? Its not part of the SonicWall Firewall connector.

**Review by v-jayakal — CHANGES_REQUESTED** [review]

**Comment by tijuc** [comment]
Hello

I had a query on Device Product value

Can I have device product value as below in the json file?

where DeviceProduct == \"firewall\",\"TZ 670\",\"NSv 270\",\"TZ 570\",\"TZ 270\",\"TZ 370\",\"TZ 470\",\"NSa 2700\",\"NSv 470\",\"NSv 870\""


Below is what I had currently in the json file but we have multiple Firewall models and hence DeviceProduct values. :

    "graphQueries": [
        {
            "metricName": "Total data received",
            "legend": "SonicWall",
            "baseQuery": "CommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\""
        }
    ],
    "sampleQueries": [
        {
            "description" : "All logs",
            "query": "\nCommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n\n            | sort by TimeGenerated"
        },
        {
            "description" : "Summarize by destination IP and port",
            "query": "\nCommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n\n            | summarize count() by DestinationIP, DestinationPort\n            | sort by TimeGenerated"
        }

    ],
    "dataTypes": [
        {
            "name": "CommonSecurityLog (SonicWall)",
            "lastDataReceivedQuery": "CommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n| summarize Time = max(TimeGenerated)\n| where isnotempty(Time)"
        }
    ],
    "connectivityCriterias": [
        {
            "type": "IsConnectedQuery",
            "value": [
                "CommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n| summarize LastLogReceived = max(TimeGenerated)\n| project IsConnected = LastLogReceived > ago(30d)"
            ]
        }
    ],

Regards,
Tiju

**Comment by v-jayakal** [comment]
@tijuc , If multiple products are to be queried, you can use "has_any" operator.
Sample query with list of Device Products: 
CommonSecurityLog
| where DeviceVendor == "SonicWall"
| where DeviceProduct has_any ("firewall","TZ 670","NSv 270","TZ 570","TZ 270","TZ 370","TZ 470","NSa 2700","NSv 470","NSv 870")

Condition :
| where DeviceProduct has_any ( \"firewall\",\"TZ 670\",\"NSv 270\",\"TZ 570\",\"TZ 270\",\"TZ 370\",\"TZ 470\",\"NSa 2700\",\"NSv 470\",\"NSv 870\")

File: DataConnectors/SonicWall/SonicwallFirewall.JSON [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    ],
+    "sampleQueries": [
+        {
+            "description" : "All logs",
+            "query": "\nCommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n\n            | sort by TimeGenerated"
+        },
+        {
+            "description" : "Summarize by destination IP and port",
+            "query": "\nCommonSecurityLog\n| where DeviceVendor == \"SonicWall\"\n| where DeviceProduct == \"firewall\"\n\n            | summarize count() by DestinationIP, DestinationPort\n            | sort by TimeGenerated"
```
[the remark this answers is not in the document]
Comment (v-jayakal) [reply DataConnectors/SonicWall/SonicwallFirewall.JSON]: This query has modified to have "count() by dst, proto\n            | sort by TimeGenerated" - There are no columns by name dst and proto in CommonSecurityLog. 
The earlier query has DestinationIP and DestinationPort - These 2 columns are correct, The problem is the summarize query return only 2 columns DestinationIP and DestinationPort. But the query has sorting by "TimeGenerated" which is not returned since result has only 2 columns DestinationIP  and DestinationPort, hence the results cannot be sorted by a column "TimeGenerated" which is not returned. 

If you want to have the results to be sorted by TimeGenerated, you have to include that column in summarize.
Working Sample Query that returns 3 columns DestinationIP, DestinationPort and TimeGenerated: 
CommonSecurityLog | where DeviceVendor == "SonicWall" | where DeviceProduct == "firewall"             
| summarize count() by DestinationIP, DestinationPort , TimeGenerated        
   | sort by TimeGenerated
-- Since TimeGenerated is returned, results can be sorted by TimeGenerated.

**Comment by tijuc** [comment]
Thankyou @v-jayakal<mailto:notifications@github.com> I have submitted the changes. Please review.

Regards,
Tiju

**Review by v-jayakal — CHANGES_REQUESTED** [review]
1. Remove "Doc_Template_CEF_Connector.md" from PR.
2. All queries are in incorrect (syntax). Attaching the file with correction to queries. Please check and update if you find it as expected.
3. Also, query for 'Show all dropped traffic from the SonicWall Firewall' uses fw_action in condition, but CommonSecurityLog does not have fw_action as field. I could find the value in AdditionalExtensions field, modified the query to check for fw_action='drop' in AdditionalExtensions field. Please check and correct it is not the expected query.
[SonicwallFirewall-update.zip](https://github.com/Azure/Azure-Sentinel/files/5634219/SonicwallFirewall-update.zip)

**Review by tijuc — COMMENTED** [review]
Added all teh changes as suggested during review

**Comment by v-jayakal** [comment]
> Added all teh changes as suggested during review

Hey, do not remove the file Doc_Template_CEF_Connector.md. Please DO NOT include it in your PR as it is not related to the connector changes you are doing. The file should not be part of your PR.

**Comment by tijuc** [comment]
Will reopen another one
